Updates behavior of plugin in several ways:

rjuang opened the pull request:
- An "Inject" icon appears for methods annotated with @Provides containing more than 1 argument.
This allows users to click the icon to find where the given type was injected from.
- An "Inject" icon no longer appears for methods annotated with @Inject having no arguments.
No dependencies are injected into empty constructors annotated with @Inject. Those only serve as providers.
- Fixes searching for injections where there is no corresponding @Provides but there exists a constructor
denoted with @Inject.
- When searching a @Provides statement for where it is injected, results are fixed to include injections in the parameter list of other @Provides statements.

Details on changes:

- Updates Decider class to be cleaner and take in the PsiElement so to enable flexibility of deriving additional information from the class later.
- Updates InjectionLineMakerProvider so that constructors with @Inject annotation must have at least 1 argument for the inject icon to appear.
- Updates InjectionLineMakerProvider so that methods with @Provides annotation and at least 1 argument have an inject icon appear.
- Updates ConstructorInjectToProvidesHandler to search for constructors with @Inject annotation first. If none exists, continues with usage search
to find injection source.
- This patch updates ProvidesMethodDecider such that methods annotated with @Provides are also included in the search for injections.

AlecStrong commented on src/com/squareup/ideaplugin/dagger/ShowUsagesAction.java:
```diff
@@ -72,34 +56,13 @@
-import com.intellij.usages.impl.UsageViewManagerImpl;
-import com.intellij.usages.impl.UsageViewTreeModelBuilder;
+import com.intellij.usages.*;
+import com.intellij.usages.impl.*;
```
make sure you're formatting with the [square code style](https://github.com/square/java-code-styles) so you dont get these star imports

AlecStrong commented on src/com/squareup/ideaplugin/dagger/Decider.java:
```diff
@@ -33,20 +34,21 @@ public ProvidesMethodDecider(PsiMethod psiMethod) {
     }
 
     @Override public boolean shouldShow(UsageTarget target, Usage usage) {
+
```
nit: revert

AlecStrong commented on src/com/squareup/ideaplugin/dagger/Decider.java:
```diff
@@ -33,20 +34,21 @@ public ProvidesMethodDecider(PsiMethod psiMethod) {
-      if (field != null //
-          && PsiConsultantImpl.hasAnnotation(field, CLASS_INJECT) //
+      if (field != null
+          && PsiConsultantImpl.hasAnnotation(field, CLASS_INJECT)
```
nit: revert (not super important but keeps the diff clean)

AlecStrong commented on src/com/squareup/ideaplugin/dagger/handler/ConstructorInjectToProvidesHandler.java:
```diff
@@ -45,8 +42,36 @@
+
+  private boolean navigateToConstructorIfProvider(PsiParameter psiParameter) {
+    PsiTypeElement declaringTypeElement = psiParameter.getTypeElement();
+    PsiClass classElement = JavaPsiFacade.getInstance(psiParameter.getProject()).findClass(
```
I believe you can do this with `(PsiClass) psiParameter.getTypeElement().reference.resolve()` instead of using `JavaPsiFacade` which is pretty expensive. I could be wrong though I haven't tried it for myself.

rjuang replied:
Just tried it with psiParameter and couldn't get it to work. I tried some other various methods too and could not get them to work.

AlecStrong replied:
coolcool :+1: thanks for trying JavaPsiFacade it is!

AlecStrong commented on src/com/squareup/ideaplugin/dagger/handler/ConstructorInjectToProvidesHandler.java:
```diff
@@ -45,8 +42,36 @@
+            declaringTypeElement.getResolveScope());
+    for (PsiMethod method : classElement.getConstructors()) {
+      if (PsiConsultantImpl.hasAnnotation(method, CLASS_INJECT)) {
+        if (navigateToElement(method)) {
```
`&&` this into the above if statement

AlecStrong commented on src/com/squareup/ideaplugin/dagger/handler/ConstructorInjectToProvidesHandler.java:
```diff
@@ -45,8 +42,36 @@
-        PsiConsultantImpl.checkForLazyOrProvider(psiParameter), new RelativePoint(mouseEvent),
-        PsiUtilBase.findEditor(psiParameter), MAX_USAGES);
+    // Check to see if class type of psiParameter has constructor with @Inject. Otherwise, proceed.
+    if (!navigateToConstructorIfProvider(psiParameter)) {
```
since `navigateToConstructorIfProvider` potentially performs a UI step I would inverse the boolean and break early. 

```java
if (navigateToConstructorIfProvider(psiParameter)) {
  return;
}
new ShowUsagesAction(...).startFindUsages(...)
```

AlecStrong commented:
Oh yea it also looks like you need to rebase against master since there are conflicts

rjuang commented:
Is it okay to have the nits addressed in a new CL?

rjuang commented:
Alright. I've squashed all my changes into one commit to cleanup the change history. Also removed the jar file since you probably want to generate the jar file and upload yourself.

rjuang commented:
And just spotted a NPE. Added fixed and amended patch.

AlecStrong commented:
Thanks!
